lfs/http.go

```diff
@@ -73,104 +154,139 @@ func checkRedirect(req *http.Request, via []*http.Request) error {
+type countingReadCloser struct {
+	Count    int
+	request  *http.Request
+	response *http.Response
+	io.ReadCloser
 }
 
-func (r *countingBody) Close() error {
-	if r.Direction == countingUpload {
-		fmt.Fprintf(os.Stderr, "* uploaded %d bytes\n", r.Size)
-	} else {
-		fmt.Fprintf(os.Stderr, "* downloaded %d bytes\n", r.Size)
+func (c *countingReadCloser) Read(b []byte) (int, error) {
+	n, err := c.ReadCloser.Read(b)
+	if err != nil && err != io.EOF {
+		return n, err
 	}
-	return r.ReadCloser.Close()
-}
 
-func newCountedResponse(res *http.Response) *countingBody {
-	return &countingBody{countingDownload, 0, res.Body}
-}
+	c.Count += n
 
-func newCountedRequest(req *http.Request) *countingBody {
-	return &countingBody{countingUpload, 0, req.Body}
-}
+	if Config.isTracingHttp {
+		contentType := ""
+		if c.response != nil { // Response, only print certain kinds of data
+			contentType = strings.ToLower(strings.SplitN(c.response.Header.Get("Content-Type"), ";", 2)[0])
+		} else {
+			contentType = strings.ToLower(strings.SplitN(c.request.Header.Get("Content-Type"), ";", 2)[0])
+		}
 
-type tracedBody struct {
-	io.ReadCloser
-}
+		for _, tracedType := range tracedTypes {
+			if strings.Contains(contentType, tracedType) {
+				fmt.Fprint(os.Stderr, string(b[0:n]))
+			}
+		}
+	}
 
-func (r *tracedBody) Read(p []byte) (int, error) {
-	n, err := r.ReadCloser.Read(p)
-	fmt.Fprintf(os.Stderr, "%s\n", string(p[0:n]))
+	if err == io.EOF && Config.isLoggingStats {
+		// This transfer is done, we're checking it this way so we can also
+		// catch transfers where the caller forgets to Close() the Body.
+		if c.response != nil {
+			transfersLock.Lock()
+			if transfer, ok := transfers[c.response]; ok {
+				transfer.responseStats.BodySize = c.Count
+				transfer.responseStats.Stop = time.Now()
+			}
+			transfersLock.Unlock()
+		}
+	}
 	return n, err
 }
 
-func (r *tracedBody) Close() error {
-	return r.ReadCloser.Close()
+// LogHttpStats is intended to be called after all HTTP operations for the
+// commmand have finished. It dumps k/v logs, one line per transfer into
+// a log file with the current timestamp.
+func LogHttpStats() {
+	if !Config.isLoggingStats {
+		return
+	}
+
+	file, err := statsLogFile()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error logging http stats: %s\n", err)
+		return
+	}
+
+	fmt.Fprintf(file, "concurrent=%d batch=%v time=%d version=%s\n", Config.ConcurrentTransfers(), Config.BatchTransfer(), time.Now().Unix(), Version)
+
+	for key, responses := range transferBuckets {
+		for _, response := range responses {
+			stats := transfers[response]
+			fmt.Fprintf(file, "key=%s reqheader=%d reqbody=%d resheader=%d resbody=%d restime=%d status=%d url=%s\n",
+				key,
+				stats.requestStats.HeaderSize,
+				stats.requestStats.BodySize,
+				stats.responseStats.HeaderSize,
+				stats.responseStats.BodySize,
+				stats.responseStats.Stop.Sub(stats.responseStats.Start).Nanoseconds(),
+				response.StatusCode,
+				stats.url)
```

Comment:
There's no need to track this `stats.url` property since you can get it from `response.Request.URL.String()`.